support-frontend/assets/pages/[countryGroupId]/components/formOnSubmit.ts

```diff
@@ -0,0 +1,204 @@
+	extractPersonalDataFromForm,
+} from '../checkout/helpers/formDataExtractors';
+import { setThankYouOrder } from '../checkout/helpers/sessionStorage';
+import type { PaymentMethod } from './paymentFields';
+import { FormSubmissionError } from './paymentFields';
+import type { ProcessPaymentResponse } from './retryPaymentStatus';
+import { processPaymentWithRetries } from './retryPaymentStatus';
+
+type CreateSubscriptionResponse = StatusResponse & {
+	userType: UserType;
+};
+
+export const submitForm = async ({
+	geoId,
+	productKey,
+	ratePlanKey,
+	formData,
+	paymentMethod,
+	paymentFields,
+	productFields,
+	hasDeliveryAddress,
+	abParticipations,
+	promotion,
+	contributionAmount,
+}: {
+	geoId: GeoId;
+	productKey: ActiveProductKey;
+	ratePlanKey: string;
+	formData: FormData;
+	paymentMethod: PaymentMethod;
+	paymentFields: RegularPaymentFields;
+	productFields: ProductFields;
+	hasDeliveryAddress: boolean;
+	abParticipations: Participations;
+	promotion: Promotion | undefined;
+	contributionAmount: number | undefined;
+}) => {
+	const personalData = extractPersonalDataFromForm(formData);
+	const { billingAddress, deliveryAddress } = hasDeliveryAddress
+		? extractDeliverableAddressDataFromForm(formData)
+		: extractNonDeliverableAddressDataFromForm(formData);
+
+	const ophanIds = getOphanIds();
+	const referrerAcquisitionData = {
+		...getReferrerAcquisitionData(),
+		labels: ['generic-checkout'], // Shall we get rid of this now?
+	};
+
+	const firstDeliveryDate =
+		productKey === 'TierThree'
+			? formatMachineDate(getTierThreeDeliveryDate())
+			: null;
+
+	const promoCode = promotion?.promoCode;
+	const appliedPromotion =
+		promoCode !== undefined
+			? {
+					promoCode,
+					countryGroupId: geoId,
+			  }
+			: undefined;
+	const supportAbTests = getSupportAbTests(abParticipations);
+
+	const createSupportWorkersRequest: RegularPaymentRequest = {
+		...personalData,
+		billingAddress,
+		deliveryAddress,
+		firstDeliveryDate,
+		paymentFields,
+		appliedPromotion,
+		ophanIds,
+		referrerAcquisitionData,
+		product: productFields,
+		supportAbTests,
+		debugInfo: '',
+	};
+
+	const createResponse = await createNewSupportWorkersExecution(
+		createSupportWorkersRequest,
+	);
+	const { userType, processPaymentResponse } = await pollForStatusOfExecution(
+		createResponse,
+	);
+
+	if (
+		processPaymentResponse.status === 'success' ||
+		processPaymentResponse.status === 'pending'
+	) {
+		goToThankYouPage(
+			productKey,
+			ratePlanKey,
+			promoCode,
+			userType,
+			contributionAmount,
+			personalData,
+			paymentMethod,
+			processPaymentResponse.status,
+			geoId,
+		);
+	} else {
+		console.error(
+			'processPaymentResponse error:',
+			processPaymentResponse.failureReason,
+		);
+		throw new FormSubmissionError(
+			'Sorry, something went wrong.',
+			appropriateErrorMessage(
+				processPaymentResponse.failureReason ?? 'unknown',
+			),
+		);
+	}
+};
+
+const createNewSupportWorkersExecution = async (
```

Comment:
Would it make sense to move these methods around creating the support workers request and polling for the status to a separate file?